File: 0967-minimum-falling-path-sum/0967-minimum-falling-path-sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& mat) {
        int n = mat.size();

        for (int row = 1; row < n; row++) {
            for (int col = 0; col < n; col++) {
                int mini = mat[row - 1][col];
                if (col - 1 >= 0) {
                    mini = min(mini, mat[row - 1][col - 1]);
                }
                if (col + 1 < n) {
                    mini = min(mini, mat[row - 1][col + 1]);
                }

                mat[row][col] += mini;
            }
        }

        int res = mat[n - 1][0];
        for (int col = 1; col < n; col++) {
            res = min(res, mat[n - 1][col]);
        }

        return res;
    }
};

File: 0967-minimum-falling-path-sum/0967-minimum-falling-path-sum.py
class Solution:
    def minFallingPathSum(self, mat: List[List[int]]) -> int:
        n = len(mat)
        for i in range(n - 2, -1, -1):
            for j in range(n - 1, -1, -1):
                if j == n - 1:
                    mat[i][j] += min(mat[i + 1][j], mat[i + 1][j - 1])
                elif j == 0:
                    mat[i][j] += min(mat[i + 1][j], mat[i + 1][j + 1])
                else:
                    mat[i][j] += min(
                        mat[i + 1][j], mat[i + 1][j + 1], mat[i + 1][j - 1]
                    )
        return min(mat[0])
